core/serix/map_encode.go

```diff
@@ -0,0 +1,259 @@
+	case reflect.Array:
+		sliceValue := sliceFromArray(value)
+		sliceValueType := sliceValue.Type()
+		return api.mapEncodeSlice(ctx, sliceValue, sliceValueType, ts, opts)
+	case reflect.Interface:
+		return api.mapEncodeInterface(ctx, value, valueType, ts, opts)
+	case reflect.String:
+		str := value.String()
+		if !utf8.ValidString(str) {
+			return nil, ErrNonUTF8String
+		}
+		return value.String(), nil
+	case reflect.Bool:
+		return value.Bool(), nil
+	case reflect.Int8, reflect.Int16, reflect.Int32:
+		return value.Int(), nil
+	case reflect.Int64:
+		return strconv.FormatInt(value.Int(), 10), nil
+	case reflect.Uint8, reflect.Uint16, reflect.Uint32:
+		return value.Uint(), nil
+	case reflect.Uint64:
+		return strconv.FormatUint(value.Uint(), 10), nil
+	case reflect.Float32, reflect.Float64:
+		return strconv.FormatFloat(value.Float(), 'E', -1, 64), nil
+	default:
+	}
+	return nil, errors.Errorf("can't encode: unsupported type %T", valueI)
+}
+
+func (api *API) mapEncodeInterface(
+	ctx context.Context, value reflect.Value, valueType reflect.Type, ts TypeSettings, opts *options,
+) (any, error) {
+	elemValue := value.Elem()
+	if !elemValue.IsValid() {
+		return nil, errors.Errorf("can't serialize interface %s it must have underlying value", valueType)
+	}
+
+	registry := api.getInterfaceObjects(valueType)
+	if registry == nil {
+		return nil, errors.Errorf("interface %s isn't registered", valueType)
+	}
+
+	elemType := elemValue.Type()
+	if _, exists := registry.fromTypeToCode[elemType]; !exists {
+		return nil, errors.Errorf("underlying type %s hasn't been registered for interface type %s",
+			elemType, valueType)
+	}
+
+	elemTypeSettings, _ := api.getTypeSettings(elemType)
+
+	ele, err := api.mapEncode(ctx, elemValue, elemTypeSettings, opts)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to encode interface element %s", elemType)
+	}
+	return ele, nil
+}
+
+func (api *API) mapEncodeStruct(
+	ctx context.Context, value reflect.Value, valueI interface{}, valueType reflect.Type, ts TypeSettings, opts *options,
+) (any, error) {
+	if valueTime, ok := valueI.(time.Time); ok {
+		return valueTime.Format(time.RFC3339Nano), nil
+	}
+
+	obj := orderedmap.New()
+	if ts.ObjectType() != nil {
+		obj.Set(mapTypeKeyName, ts.ObjectType())
+	}
+	if err := api.mapEncodeStructFields(ctx, obj, value, valueType, opts); err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return obj, nil
+}
+
+func (api *API) mapEncodeStructFields(
+	ctx context.Context, obj *orderedmap.OrderedMap, value reflect.Value, valueType reflect.Type, opts *options,
+) error {
+	structFields, err := parseStructType(valueType)
+	if err != nil {
+		return errors.Wrapf(err, "can't parse struct type %s", valueType)
+	}
+
+	for _, sField := range structFields {
+		fieldValue := value.Field(sField.index)
+		if sField.isEmbeddedStruct && !sField.settings.nest {
+			fieldType := sField.fType
+			if fieldValue.Kind() == reflect.Ptr {
+				if fieldValue.IsNil() {
```

Comment:
is it allowed to skip here, even if we don't check omitempty or isoptional?

Reply:
If the inner embedded struct is a pointer and is nil, then there are no fields of it to be written to the map anyway, so skipping here should be fine.